design/rest-api-guideline.md

```diff
@@ -0,0 +1,388 @@
+# REST API Guideline
+
+* **Status**: Draft
+* **Author**: [pedroigor](https://github.com/pedroigor), [abstractj](https://github.com/abstractj), [ssilvert](https://github.com/ssilvert)
+* **JIRA**: [KEYCLOAK-9344](https://issues.jboss.org/browse/KEYCLOAK-9344)
+
+## Abstract
+
+The Keycloak REST API Guideline provides a set of design principles and practices that should be considered by developers when designing, implementing and exposing a RESTful API. To provide the best experience for developers consuming Keycloak REST APIs, it's important API developers to follow to the principles and practices defined herein, making these APIs intuitive and easy to use.
+
+The goals of the guideline are:
+
+* Define the [overall principles](#principles) that should be taken into account when designing REST APIs for Keycloak
+
+* Define common practices and patterns for all Keycloak REST APIs such as:
+    * [Filtering](#filtering)
+    * [Pagination](#pagination)
+    * [Error Handling](#error-handling)
+    * [Documentation](#documentation)
+
+* Define common practices for [documentation](#documentation)
+
+Although the guideline is based on the REST architectural style, it does not try to strictly adhere to RESTful constraints. The principles covered by the guideline are based on practical experiences in the development of REST APIs in Keycloak while still conforming with most of the constraints of REST architectures.
+
+**Table of Contents**
+
+
+  - [Principles](#principles)
+    - [Level of Service Maturity](#level-of-service-maturity)
+    - [Resource Path](#resource-path)
+      - [Group](#group)
+      - [Versioning](#versioning)
+      - [Extensions](#extensions)
+      - [Technology Preview](#technology-preview)
+      - [Store Resources](#store-resources)
+      - [Controller Resources](#controller-resources)
+    - [Resource Representation](#resource-representation)
+    - [HTTP Status Codes](#http-status-codes)
+    - [HTTP Verbs](#http-verbs)
+      - [GET](#get)
+      - [POST](#post)
+      - [PUT](#put)
+      - [DELETE](#delete)
+  - [Practices](#practices)
+    - [Error Handling](#error-handling)
+    - [Filtering](#filtering)
+    - [Pagination](#pagination)
+    - [Documentation](#documentation)
+  - [Examples](#examples)
+
+## Principles
+
+This section defines the general principles that should be considered by API developers, regardless of the practices defined through this document.
+
+### Level of Service Maturity
+
+Based on [Richardson Maturity Model](https://martinfowler.com/articles/richardsonMaturityModel.html), developers MUST adhere to the maturity Level 2, where interactions with the API endpoints are based on the usage of HTTP Verbs and status codes to communicate status and errors.
+
+### Resource Path
```

Comment:
On one side I do like this, but there are at least two issues with it:

* admin endpoints are special and folks will often want to only expose these internally, both console and endpoints. With this there would be '/admin' for the console, with '/apis/admin/' for endpoints, which both have to be blocked for external traffic (or bound to internal IP/separate port).
* At some point I would like to be able to support virtual names for realms. For example allow myrealm.company.com to map to myrealm. That would require the realm to be the first part in the URL.

I suggest:

* '/apis/{API_GROUP}' - for APIs that do not operate on a specific realm (I don't think we have or will have any though, so probably not needed)
* '/realms/{REALM}/apis/{API_GROUP} - for APIs that operate on a specific realm (I believe this actually applies to all APIs)

Reply:
Well, that goes against the idea of the guideline and similar to current resource path format. I was expecting this discussion as this is one of the most critical points of the guideline.

Firstly, I don't think that `/apis/{API_GROUP}' is not needed. If you look at the example section, you'll see how we would use the format defined by this guideline with the User API. Let me put it here:

|Resource Path|GET|POST|PUT|DELETE|
|---|---|---|---|---|
|/apis/users/v1/realms/acme|List all users in the realm `acme`|Creates an user in the realm `acme`|Not Supported|Not Supported|
|/apis/users/v1/123|Gets a representation for user with id `123`|Not Supported|Updates the user with id `123`|Deletes the user with id `123`|

I really think that this format is much more descriptive, concise and aligned with our functionalities. It provides a great separation and  a versioning schema on a per-api basis. More justifications you'll find in the `Group` section of the document.

Also, note that you don't actually need the realm in the path when managing a specific user. You can perform all operations for a particular user just based on its ID.

The same goes with `/apis/accounts`, why you need the realm if you are operating in the context of a specific user ? The realm is implicit.

Regarding the protection of the APIs part of the admin logical group. I don't think the path format impacts so much on how you would protect these APIs. You could, for instance, blacklist APIs and whitelist only those you want to expose.

Regarding the virtual names. You kind of push me to the corner with this one :) I wasn't aware of this requirement and is it so special to force us to use a specific path format and ignore other options (eventually better)?

Taking the path we use for login as an example, today we have `/realms/{realm}/protocol/openid-connect/auth`. With the format I'm proposing the path would be `/apis/auth/v1/{realm}/openid-connect` (ok, we don't have a plural in the api group name here :)). Where `auth` is the API group and the version schema is specific for the Authentication API which can change in order to support, for instance, future and breaking versions of openid-connect.

Reply:
I was not expecting a URL scheme to be defined as part of the REST guidelines. That is an additional bonus, but it does introduce a lot more to discuss ;)

I see absolutely no benefits to putting the realm in the place that you have. It's confusing and wrong. As a user I do not know when to include it and when to not include it. I can't simply prefix URLs including the realm as is possible today. I can't add support for virtual host for realms (sorry, but this is something I want to at least be able to add in the future).

Also, we always need to know what realm we are operating under. A request to /apis/users/v1/123 needs to know the realm in order to know how to verify the bearer token. Further, sure today a user id is unique as it is a uuid. That is not a given that it will be in the future. We may even have users loaded from different locations for different realms.

As a realm is our tenant it should always be specified and should be a top-level thing.

We can continue the discussion around this if you want, but you're going to have a very hard time to convince me here.

Reply:
OK, you have a decision then and I won't continue the discussion. I have a different opinion, I don't think is wrong (can give you well-known examples), and I do see benefits with this approach.

I'll update the documentation accordingly.